Initial Implementation

shore opened the pull request:
Action to test product versions for embedded non-printing and whitespace characters.

* Sample [job](https://github.com/hashicorp/actions-validate-product-version/actions/runs/5180013726/jobs/9333650147#step:3:33) showing error output when version strings don't match
* Sample [job](https://github.com/hashicorp/actions-validate-product-version/actions/runs/5261799593) (success)
* Sample [job](https://github.com/hashicorp/actions-validate-product-version/actions/runs/5261799599) testing latest releases of products (success)

dekimsey reviewed (changes requested):
Nice work! Other than the boundary false-positive, this looks really good!

shore commented on .github/workflows/test-latest-products.yml:
```diff
@@ -0,0 +1,63 @@
+      fail-fast: false # run all jobs so we see all failures as soon as possible
+      matrix:
+        tool:
+          - consul
+          - nomad
+          - packer
+          - terraform
+          - vault
```
Turns out we have quite a few things going through CRT that won't work, so I'm going to have to make the action only apply to cert products.  I'm not sure how many should be in this workflow's list, though.

shore commented:
Ok, I've rewritten most of this action.  It's more robust, easier to test, and adding support for more products to test is simpler.  I've updated the links in the PR description to refer to recent test runs.

shore commented on test-versions:
```diff
@@ -0,0 +1,124 @@
+        log "==> Output (with special characters replaced by '${REPL_CHAR}')"
+        log "$output"
+        return 1
+    fi
```
I've omitted Boundary.  We talked offline, but Boundary will not be supported with it's current output format.

dekimsey commented on .github/workflows/test-latest-products.yml:
```diff
@@ -0,0 +1,80 @@
+
+      - name: Load Supported Product Build Matrix
+        id: load-matrix
+        run: echo "matrix=$(tr '\n' ' ' < supported_builds.json | tr -s ' ' )" >> "$GITHUB_OUTPUT"
```
If I'm reading this right, this slurps in a json file and removes newlines? What do you think of leveraging jq's compact mode to do the work?

```suggestion
        run: echo "matrix=$(jq --compact-output < supported_builds.json)" >> "$GITHUB_OUTPUT"
```

dekimsey commented on .github/workflows/test-latest-products.yml:
```diff
@@ -0,0 +1,80 @@
+    outputs:
+      supported: ${{ steps.load-matrix.outputs.matrix }}
+    steps:
+      - uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab
```
```suggestion
      - uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab # v3.5.2
```

dekimsey commented on .github/workflows/test-latest-products.yml:
```diff
@@ -0,0 +1,80 @@
+      matrix:
+        include: ${{ fromJSON( needs.load-matrix.outputs.supported ) }}
+    steps:
+      - uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab
```
```suggestion
      - uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab # v3.5.2
```

dekimsey commented on action.yml:
```diff
@@ -0,0 +1,58 @@
+
+        if [ "$supported" != 'true' ] ; then
+          echo "::error ${tool} is not supported by this action." 1>&2
+          [ -n "${{ inputs.soft_fail }}" ] && exit 0 # exit 1 if soft_fail unset
```
Instead of aborting (that is what's happening here right?) the test on unsupported products, is it feasible perhaps to take a more TDD approach and instead to test for failure? That way when a product starts succeeding we report the unexpected change so are hardcoded list gets updated.

I dunno, just a thought. Kinda depends on how many unsupported products we think we will catch.

shore replied:
Reporting how?  I don't want to try to pull slack notification (or secrets) into this, so anything that doesn't fail the workflow won't get seen -- the output is collapsed, the workflow gets a green check, and the release process continues.

We don't actually need the hardcoded list to be updated, as product teams that want this can call the action directly from their build.yml workflow.  They can opt-in with no action on our side.

dekimsey reviewed (approved):
Very cool! The supporter products index is an interesting approach. It very much makes sense given the constraints. It's better than my own conclusion of, "oh well, no can do". Hopefully with some prodding we can get more products supported.

shore commented:
Updated pinned versions, force-push to squash that into the original commit.
